@nicola-wiseman the test fails if you set log="both"

```
======================================================================
ERROR: test_post_ensemble (__main__.TestRegressors.test_post_ensemble)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/phyto-2/Abil/tests/test.py", line 34, in test_post_ensemble
    m.train(model="xgb", log="both")
  File "/home/phyto-2/Abil/abil/tune.py", line 259, in train
    logger.info(f"best fit: {reg_grid_search.best_estimator_.named_steps['estimator'].transformation_}")
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'TransformedTargetRegressor' object has no attribute 'named_steps'

======================================================================
ERROR: test_post_ensemble (__main__.Test2Phase.test_post_ensemble)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/phyto-2/Abil/tests/test.py", line 128, in test_post_ensemble
    m.train(model="xgb", log="both")
  File "/home/phyto-2/Abil/abil/tune.py", line 259, in train
    logger.info(f"best fit: {reg_grid_search.best_estimator_.named_steps['estimator'].transformation_}")
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'TransformedTargetRegressor' object has no attribute 'named_steps'

----------------------------------------------------------------------
Ran 6 tests in 5.714s

FAILED (errors=2)
```

It is OK if these fail (since Log will give a different number):

```
        self.assertAlmostEqual(integrated_total_poc_mean, 2.55e17, delta=1e15)
        self.assertAlmostEqual(integrated_total_poc_ci95_UL, 3.92e+17, delta=1e15)
        self.assertAlmostEqual(integrated_total_poc_ci95_LL, 1.44e+17, delta=1e15)
```

But the bug seems to be related to something else...

--- a/abil/tune.py
+++ b/abil/tune.py
@@ -256,7 +256,14 @@ class ModelTuner:
                 reg_grid_search = reg.transformed_fit(X_train, y, log, self.model_config['predictors'].copy())
             
             if log == 'both':
-                logger.info(f"best fit: {reg_grid_search.best_estimator_.named_steps['estimator'].transformation_}")
+                best = reg_grid_search.best_estimator_
+                # If it's a pipeline → extract "estimator"
+                if hasattr(best, "named_steps"):
+                    ttr = best.named_steps["estimator"]
+                else:
+                    # Bare Transformed Target Regressor
+                    ttr = best
+                logger.info(f"best fit: {ttr.transformation_}")
             m2 = reg_grid_search.best_estimator_
 
 

--- a/tests/test.py
+++ b/tests/test.py
@@ -99,9 +99,9 @@ class TestRegressors(unittest.TestCase):
             os.path.join("tests", "ModelOutput", "regressor", "posts", "integrated_totals", "ens_integrated_totals_ci95_LL_poc.csv")
         )["total"].iat[0]
 
-        self.assertAlmostEqual(integrated_total_poc_mean, 1.013e17, delta=1e15)
-        self.assertAlmostEqual(integrated_total_poc_ci95_UL, 1.95e17, delta=1e15)
-        self.assertAlmostEqual(integrated_total_poc_ci95_LL, 2.46e16, delta=1e15)
+        self.assertAlmostEqual(integrated_total_poc_mean, 2.301e17, delta=1e15)
+        self.assertAlmostEqual(integrated_total_poc_ci95_UL, 3.67e17, delta=1e15)
+        self.assertAlmostEqual(integrated_total_poc_ci95_LL, 3.752e16, delta=1e15)
 
 
 class Test2Phase(unittest.TestCase):
@@ -189,9 +189,9 @@ class Test2Phase(unittest.TestCase):
             os.path.join("tests", "ModelOutput", "2-phase", "posts", "integrated_totals", "ens_integrated_totals_ci95_LL_poc.csv")
         )["total"].iat[0]
     
-        self.assertAlmostEqual(integrated_total_poc_mean, 2.55e17, delta=1e15)
-        self.assertAlmostEqual(integrated_total_poc_ci95_UL, 3.92e+17, delta=1e15)
-        self.assertAlmostEqual(integrated_total_poc_ci95_LL, 1.44e+17, delta=1e15)
+        self.assertAlmostEqual(integrated_total_poc_mean, 3.27e17, delta=1e15)
+        self.assertAlmostEqual(integrated_total_poc_ci95_UL, 5.267e+17, delta=1e15)
+        self.assertAlmostEqual(integrated_total_poc_ci95_LL, 1.795e+17, delta=1e15)
 
 
 

